### PR TITLE
Add task list selection to Tasks panel

### PR DESCRIPTION
**Pull request by tttzach**
Screenshot 2020-07-26 at 6 50 56 PM
Screenshot 2020-07-26 at 6 51 20 PM
Screenshot 2020-07-26 at 6 51 05 PM

### CODE REVIEW

File: src/main/webapp/populateData.js [hunk 1]
```diff
@@ -90,12 +90,24 @@ function populateGmail() {
...
-  const tasksContainer = document.querySelector('#tasks');
+  let fetchFrom;
+  const select = document.querySelector('#panel__task-list-select');
+  // Cast from HTMLOptionsCollection to Array
+  const options = Array(...select.options);
 
-  // Get list of tasks from user's Tasks account
-  // and display the task titles from all task lists on the screen
-  fetch('/tasks')
+  if (options.length == 0) {
```
Comment (austinteshuba) [inline src/main/webapp/populateData.js:98]: Prefer `===` instead of `==` (though I can't think of a case where it matters here, this is more for consistency)

File: src/main/webapp/populateData.js [hunk 2]
```diff
@@ -90,12 +90,24 @@ function populateGmail() {
  * Populate Tasks container with user information
  */
 function populateTasks() {
-  // Get Container for Tasks content
-  const tasksContainer = document.querySelector('#tasks');
+  let fetchFrom;
+  const select = document.querySelector('#panel__task-list-select');
```
Comment (austinteshuba) [inline src/main/webapp/populateData.js:94]: Nice BEM naming. Just an FYI, you don't have to follow this for id's - it's more for classes. Not an issue though

File: src/main/webapp/populateData.js [hunk 3]
```diff
@@ -104,6 +116,14 @@ function populateTasks() {
         return response.json();
       })
       .then((tasksResponse) => {
+        const taskListNames = tasksResponse['taskListNames'];
```
Comment (austinteshuba) [inline src/main/webapp/populateData.js:119]: nit: use `allTaskListNames` for consistency

File: src/main/webapp/populateData.js [hunk 4]
```diff
@@ -104,6 +116,14 @@ function populateTasks() {
...
+        const select = document.querySelector('#panel__task-list-select');
```
Comment (austinteshuba) [inline src/main/webapp/populateData.js:120]: Reuse the above `select` declaration

File: src/main/java/com/google/sps/servlets/TasksServlet.java [hunk 5]
```diff
@@ -141,20 +145,30 @@ public void doGet(
...
-    List<TaskList> taskLists = tasksClient.listTaskLists();
-    List<Task> tasks = getTasks(tasksClient);
+    List<TaskList> allTaskLists = tasksClient.listTaskLists();
 
     // Initialize Tasks Response
-    List<String> taskListTitles = getTaskListTitles(taskLists);
+    List<String> allTaskListTitles = getTaskListTitles(allTaskLists);
+    List<Task> tasks;
+
+    String queryString = request.getQueryString();
```
[the remark this answers is not in the document]
Comment (tttzach) [reply src/main/java/com/google/sps/servlets/TasksServlet.java:156]: I don't think I need any other parameter for this servlet so this is great, thanks!

File: src/main/webapp/index.html [hunk 6]
```diff
@@ -80,7 +80,8 @@ <h3 class="panel__section-heading">Mail</h3>
             </a>
             <h3 class="panel__section-heading">Tasks</h3>
           </div>
-          <p class="panel__section-subheading">Selected Tasks:
+          <p class="panel__section-subheading">Selected Tasks: 
+            <select name="panel__task-list-select" id="panel__task-list-select" onchange="populateTasks()" size="2" multiple></select>
```
[the remark this answers is not in the document]
Comment (tttzach) [reply src/main/webapp/index.html:84]: Yep you can but then it would scale based on how many tasklists there are and would mess up the rest of the UI on the panel. Unfortunately, size 1 was a little too small to see what was going on.

File: src/main/webapp/populateData.js [hunk 7]
```diff
@@ -265,4 +285,3 @@ function populateGo() {
         }
       });
 }
-
```
[the remark this answers is not in the document]
Comment (tttzach) [reply src/main/webapp/populateData.js:268]: It was an extra newline, GitHub doesn't show the last newline which is why it looked like there was no newline!

File: src/main/webapp/populateData.js (hunk 1 above)
Comment (tttzach) [reply src/main/webapp/populateData.js:98]: Ok sure! Consistency is always great

File: src/main/webapp/auth.js [hunk 8]
```diff
@@ -96,7 +96,8 @@ function onSignIn(googleUser) {
   // or when the browser is closed
   addCookie('accessToken', accessToken, expiryUtcTime);
 
-  handleAuthenticationState();
+  // TODO: Confirm that this is a double call
+  // handleAuthenticationState();
```
[the remark this answers is not in the document]
Comment (tttzach) [reply src/main/webapp/auth.js:100]: Oh okay that's not great then, i'll revert this. Thanks for letting me know!

File: src/main/webapp/populateData.js (hunk 2 above)
Comment (tttzach) [reply src/main/webapp/populateData.js:94]: Oh okay, i'll take off the `panel__` to keep it consistent, thanks!

File: src/main/webapp/populateData.js (hunk 3 above)
Comment (tttzach) [reply src/main/webapp/populateData.js:119]: Good idea - thanks!

File: src/main/webapp/populateData.js (hunk 4 above)
Comment (tttzach) [reply src/main/webapp/populateData.js:120]: Thanks for catching that!

File: src/main/webapp/auth.js [hunk 9]
```diff
@@ -60,7 +60,7 @@ function handleAuthenticationState() {
     populateGmail();
     populateTasks();
     populateCalendar();
-    postAndGetTaskList();
+    // postAndGetTaskList();
```
Comment (antmar) [inline src/main/webapp/auth.js:63]: Did you intend to leave this commented out? Why?

File: src/main/webapp/populateData.js [hunk 10]
```diff
@@ -90,12 +90,24 @@ function populateGmail() {
...
+  if (options.length === 0) {
+    fetchFrom = '/tasks';
+  } else {
+    const selectedOptions = [];
+    options.forEach((option) => {
+      if (option.selected) {
+        selectedOptions.push(option.value);
+      }
+    });
+    fetchFrom = '/tasks?selected=' + selectedOptions.join();
```
Comment (antmar) [inline src/main/webapp/populateData.js:107]: What does "join()" use as the default delimiter? Do you need to escape the delimiter if it is found in the task list names?

(You want to be consistent in what delimiter you use across the entire codebase.)

File: src/main/webapp/auth.js (hunk 9 above)
Comment (antmar) [inline src/main/webapp/auth.js:63]: Did you intend to leave this commented out?

File: src/main/webapp/index.html [hunk 11]
```diff
@@ -81,25 +81,26 @@ <h3 class="panel__section-heading">Mail</h3>
             </a>
             <h3 class="panel__section-heading">Tasks</h3>
           </div>
-          <p class="panel__section-subheading">Selected Tasks:
+          <p class="panel__section-subheading">Selected Tasks:
```
Comment (antmar) [inline src/main/webapp/index.html:84]: Is this selecting tasks or task lists? This text and the select id imply you are selecting tasks.

File: src/main/webapp/populateData.js [hunk 12]
```diff
@@ -104,19 +116,29 @@ function populateTasks() {
         return response.json();
       })
       .then((tasksResponse) => {
+        if (options.length === 0) {
```
Comment (antmar) [inline src/main/webapp/populateData.js:119]: Why not update the set of task lists every time?

File: src/main/webapp/populateData.js (hunk 10 above)
Comment (antmar) [inline src/main/webapp/populateData.js:107]: Nit: "selected" is a very generic parameter name. I suggest something more specific like `taskListsSelected` or just `taskLists`.

File: src/main/webapp/auth.js (hunk 9 above)
Comment (tttzach) [reply src/main/webapp/auth.js:63]: Echoing our offline conversation - this function adds tasks to a user's Tasks account on every run of the project. This change has been moved to PR #120.

Comment (tttzach) [reply src/main/webapp/auth.js:63]: Echoing our offline conversation - this function adds tasks to a user's Tasks account on every run of the project. This change has been moved to PR #120.

File: src/main/webapp/populateData.js (hunk 12 above)
Comment (tttzach) [reply src/main/webapp/populateData.js:119]: There's a double call of `handleAuthenticationState` so this is just a workaround - updating the set of task lists every time would mean there's two copies of the same task lists. Here's the [discussion](https://github.com/googleinterns/step7-2020/pull/114#discussion_r460953024) for more context.

File: src/main/webapp/index.html (hunk 11 above)
Comment (tttzach) [reply src/main/webapp/index.html:84]: Agreed, naming seems to be a common cause of confusion for tasks and task lists. Thanks for pointing that out!

File: src/main/webapp/populateData.js (hunk 10 above)
Comment (tttzach) [reply src/main/webapp/populateData.js:107]: Agreed, thanks!

Comment (tttzach) [reply src/main/webapp/populateData.js]: Good point, `join()` uses `,` however that seems to be a valid character for task list names - this seems to be the case for every key on the keyboard and I don't see why they would make any of the characters invalid either. Will bring this up on the standup today!

**Comment by tttzach** [comment]
Replying to the general comments here
@austinteshuba You're right about Empty being a task list name. The "must test edge cases mentality" pops out more for me in situations when it is not needed and disappears when it is most needed. Sorry for the confusion :/
@antmar Echoing our offline discussion - I messed up while merging conflicts and overwrote some of my code. They're back now!

File: src/main/webapp/populateData.js (hunk 10 above)
Comment (tttzach) [reply src/main/webapp/populateData.js]: `RESOLVED`

File: src/main/webapp/index.html (hunk 11 above)
Comment (tttzach) [reply src/main/webapp/index.html]: `RESOLVED`

File: src/main/webapp/auth.js (hunk 9 above)
Comment (tttzach) [reply src/main/webapp/auth.js]: `RESOLVED`

Comment (tttzach) [reply src/main/webapp/auth.js]: `RESOLVED`

File: src/main/webapp/populateData.js (hunk 4 above)
Comment (tttzach) [reply src/main/webapp/populateData.js]: `RESOLVED`

File: src/main/webapp/populateData.js (hunk 3 above)
Comment (tttzach) [reply src/main/webapp/populateData.js]: `RESOLVED`

File: src/main/webapp/populateData.js (hunk 2 above)
Comment (tttzach) [reply src/main/webapp/populateData.js]: `RESOLVED`

File: src/main/webapp/populateData.js (hunk 1 above)
Comment (tttzach) [reply src/main/webapp/populateData.js]: `RESOLVED`

File: src/main/webapp/populateData.js (hunk 7 above)
[the remark this answers is not in the document]
Comment (tttzach) [reply src/main/webapp/populateData.js:313]: `RESOLVED`

File: src/main/webapp/index.html (hunk 6 above)
[the remark this answers is not in the document]
Comment (tttzach) [reply src/main/webapp/index.html]: `RESOLVED`

File: src/main/webapp/auth.js (hunk 8 above)
[the remark this answers is not in the document]
Comment (tttzach) [reply src/main/webapp/auth.js]: `RESOLVED`

File: src/main/java/com/google/sps/servlets/TasksServlet.java (hunk 5 above)
[the remark this answers is not in the document]
Comment (tttzach) [reply src/main/java/com/google/sps/servlets/TasksServlet.java]: `RESOLVED`

File: src/main/webapp/populateData.js (hunk 10 above)
Comment (tttzach) [reply src/main/webapp/populateData.js]: From our standup today - we agreed to map task list names to task list IDs and use the task list ID as a means of identifying to exact task list to summarize key information from.

**Comment by tttzach** [comment]
Note to self - refactor to use task list IDs instead of task list names to identify task lists.

File: src/main/java/com/google/sps/servlets/TasksServlet.java [hunk 13]
```diff
@@ -147,8 +160,8 @@ public void doPost(
     }
   }
 
-  private List<String> getTaskListTitles(List<TaskList> taskLists) throws IOException {
-    return taskLists.stream().map(taskList -> taskList.getTitle()).collect(Collectors.toList());
+  private Map<String, String> getTaskListIdsAndTitles(List<TaskList> taskLists) throws IOException {
+    return taskLists.stream().collect(Collectors.toMap(TaskList::getId, TaskList::getTitle));
```
Comment (antmar) [inline src/main/java/com/google/sps/servlets/TasksServlet.java:164]: Nice use of `Collectors.toMap`

File: src/main/java/com/google/sps/servlets/TasksServlet.java [hunk 14]
```diff
@@ -79,25 +81,36 @@ public void doGet(
 
     // Get tasks from Google Tasks
     TasksClient tasksClient = tasksClientFactory.getTasksClient(googleCredential);
-    List<TaskList> taskLists = tasksClient.listTaskLists();
-    List<Task> tasks = getTasks(tasksClient);
+    List<TaskList> allTaskLists = tasksClient.listTaskLists();
 
     // Initialize Tasks Response
-    List<String> taskListTitles = getTaskListTitles(taskLists);
+    Map<String, String> allTaskListIdsAndTitles = getTaskListIdsAndTitles(allTaskLists);
```
Comment (gbuenoandrade) [inline src/main/java/com/google/sps/servlets/TasksServlet.java:87]: Nit: `allTaskListIdsToTitles` would make the map contents a bit more clear (here and everywhere else)

File: src/main/java/com/google/sps/servlets/TasksServlet.java [hunk 15]
```diff
@@ -79,25 +81,36 @@ public void doGet(
...
+    List<Task> tasks;
+
+    String taskLists = request.getParameter("taskLists");
+    if (taskLists == null) {
+      List<String> allTaskListIds =
+          allTaskListIdsAndTitles.keySet().stream().collect(Collectors.toList());
+      tasks = getTasks(tasksClient, allTaskListIds);
```
[the remark this answers is not in the document]
Comment (gbuenoandrade) [reply src/main/java/com/google/sps/servlets/TasksServlet.java:94]: +1

File: src/main/java/com/google/sps/servlets/TasksServlet.java (hunk 13 above)
Comment (tttzach) [reply src/main/java/com/google/sps/servlets/TasksServlet.java:164]: Thanks!

File: src/main/java/com/google/sps/servlets/TasksServlet.java (hunk 14 above)
Comment (tttzach) [reply src/main/java/com/google/sps/servlets/TasksServlet.java]: I agree, changed and thanks!

Comment (tttzach) [reply src/main/java/com/google/sps/servlets/TasksServlet.java]: `RESOLVED`

File: src/main/java/com/google/sps/servlets/TasksServlet.java (hunk 13 above)
Comment (tttzach) [reply src/main/java/com/google/sps/servlets/TasksServlet.java:161]: `RESOLVED`

File: src/main/java/com/google/sps/servlets/TasksServlet.java (hunk 15 above)
[the remark this answers is not in the document]
Comment (tttzach) [reply src/main/java/com/google/sps/servlets/TasksServlet.java]: I agree, the clarity is probably more useful than the code reuse. I've changed this - thanks!

[the remark this answers is not in the document]
Comment (tttzach) [reply src/main/java/com/google/sps/servlets/TasksServlet.java]: `RESOLVED`
